File: VB/Q144234/MyTreeList.vb
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes

Namespace DXSample
    Public Class MyTreeList
        Inherits TreeList
        Dim selectedRect As Rectangle = Rectangle.Empty
        Dim dragRect As Rectangle = Rectangle.Empty
        Dim startPoint As Point = Point.Empty
        Public Sub New()
            MyBase.New()
            OptionsBehavior.ShowEditorOnMouseUp = True
            OptionsSelection.EnableAppearanceFocusedCell = False
            OptionsSelection.MultiSelect = True
            OptionsView.FocusRectStyle = DrawFocusRectStyle.None
            AddHandler PaintEx, AddressOf OnPaintEx
        End Sub

        Private Sub OnPaintEx(ByVal sender As Object, ByVal e As TreeListPaintEventArgs)
            If selectedRect = Rectangle.Empty Then
                Return
            End If
            e.Cache.DrawRectangle(Pens.Blue, selectedRect)
            Using selectedBrush As New SolidBrush(Color.FromArgb(50, Color.Blue))
                e.Cache.FillRectangle(selectedBrush, selectedRect)
            End Using
        End Sub
        Protected Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
            If e.Button = MouseButtons.Left Then
                dragRect.X = e.X - SystemInformation.DragSize.Width \ 2
                dragRect.Y = e.Y - SystemInformation.DragSize.Height \ 2
                dragRect.Size = SystemInformation.DragSize
                startPoint = e.Location
                selectedRect.Location = e.Location
            End If
            MyBase.OnMouseDown(e)
        End Sub

        Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
            If e.Button = MouseButtons.Left AndAlso selectedRect <> Rectangle.Empty Then
                Selection.Clear()
                For i As Integer = 0 To VisibleNodesCount - 1
                    Dim node As TreeListNode = GetNodeByVisibleIndex(i)
                    If ViewInfo.RowsInfo(node) IsNot Nothing AndAlso ViewInfo.RowsInfo(node).Bounds.IntersectsWith(selectedRect) Then
                        node.Selected = True
                    End If
                Next i
                selectedRect = Rectangle.Empty
                dragRect = Rectangle.Empty
            End If
            MyBase.OnMouseUp(e)
            Invalidate()
        End Sub

        Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
            If e.Button = MouseButtons.Left AndAlso dragRect <> Rectangle.Empty Then
                If Not dragRect.Contains(e.Location) Then
                    If Selection.Count > 0 Then
                        Selection.Clear()
                    End If
                    Dim tmp As Rectangle = selectedRect
                    If e.X <= startPoint.X Then
                        selectedRect.X = e.X
                        selectedRect.Width += tmp.Right - selectedRect.Right
                    Else
                        selectedRect.Width = e.X - selectedRect.X
                    End If

                    If e.Y <= startPoint.Y Then
                        selectedRect.Y = e.Y
                        selectedRect.Height += tmp.Bottom - selectedRect.Bottom
                    Else
                        selectedRect.Height = e.Y - selectedRect.Y
                    End If
                    Invalidate()
                End If
            End If
            MyBase.OnMouseMove(e)
        End Sub
        Protected Overrides Sub OnMouseLeave(ByVal e As EventArgs)
            If dragRect <> Rectangle.Empty Then
                dragRect = Rectangle.Empty
                selectedRect = Rectangle.Empty
            End If
            MyBase.OnMouseLeave(e)
        End Sub
        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                RemoveHandler PaintEx, AddressOf OnPaintEx
            End If
            MyBase.Dispose(disposing)
        End Sub
    End Class
End Namespace

File: CS/Q144234/MyTreeList.cs
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;
using System;
using System.Drawing;
using System.Windows.Forms;

namespace DXSample {
    public class MyTreeList : TreeList {
        Rectangle selectedRect = Rectangle.Empty;
        Rectangle dragRect = Rectangle.Empty;
        Point startPoint = Point.Empty;
        public MyTreeList()
            : base() {
            OptionsBehavior.ShowEditorOnMouseUp = true;
            OptionsSelection.EnableAppearanceFocusedCell = false;
            OptionsSelection.MultiSelect = true;
            OptionsView.FocusRectStyle = DrawFocusRectStyle.None;
            PaintEx += OnPaintEx;
        }
        void OnPaintEx(object sender, TreeListPaintEventArgs e) {
            if(selectedRect == Rectangle.Empty) return;
            e.Cache.DrawRectangle(Pens.Blue, selectedRect);
            using(SolidBrush selectedBrush = new SolidBrush(Color.FromArgb(50, Color.Blue)))
                e.Cache.FillRectangle(selectedBrush, selectedRect);
        }
        protected override void OnMouseDown(MouseEventArgs e) {
            if(e.Button == MouseButtons.Left) {
                dragRect.X = e.X - SystemInformation.DragSize.Width / 2;
                dragRect.Y = e.Y - SystemInformation.DragSize.Height / 2;
                dragRect.Size = SystemInformation.DragSize;
                startPoint = e.Location;
                selectedRect.Location = e.Location;
            }
            base.OnMouseDown(e);
        }
        protected override void OnMouseUp(MouseEventArgs e) {
            if(e.Button == MouseButtons.Left && selectedRect != Rectangle.Empty) {
                Selection.Clear();
                for(int i = 0; i < VisibleNodesCount; i++) {
                    TreeListNode node = GetNodeByVisibleIndex(i);
                    if(ViewInfo.RowsInfo[node] != null && ViewInfo.RowsInfo[node].Bounds.IntersectsWith(selectedRect))
                        node.Selected = true;
                }
                selectedRect = Rectangle.Empty;
                dragRect = Rectangle.Empty;
            }
            base.OnMouseUp(e);
            Invalidate();
        }
        protected override void OnMouseMove(MouseEventArgs e) {
            if(e.Button == MouseButtons.Left && dragRect != Rectangle.Empty) {
                if(!dragRect.Contains(e.Location)) {
                    if(Selection.Count > 0) Selection.Clear();
                    Rectangle tmp = selectedRect;
                    if(e.X <= startPoint.X) {
                        selectedRect.X = e.X;
                        selectedRect.Width += tmp.Right - selectedRect.Right;
                    }
                    else {
                        selectedRect.Width = e.X - selectedRect.X;
                    }
                    if(e.Y <= startPoint.Y) {
                        selectedRect.Y = e.Y;
                        selectedRect.Height += tmp.Bottom - selectedRect.Bottom;
                    }
                    else {
                        selectedRect.Height = e.Y - selectedRect.Y;
                    }
                    Invalidate();
                }
            }
            base.OnMouseMove(e);
        }
        protected override void OnMouseLeave(EventArgs e) {
            if(dragRect != Rectangle.Empty) {
                dragRect = Rectangle.Empty;
                selectedRect = Rectangle.Empty;
            }
            base.OnMouseLeave(e);
        }
        protected override void Dispose(bool disposing) {
            if(disposing) {
                PaintEx -= OnPaintEx;
            }
            base.Dispose(disposing);
        }
    }
}